// @paypal/funding-components/server/util.js
"use strict";

exports.__esModule = true;
exports.buildCSP = buildCSP;
exports.getNonce = getNonce;
exports.getQuery = getQuery;
exports.getTimestamp = getTimestamp;
exports.isIE = isIE;
exports.normalizeTimestamp = normalizeTimestamp;
exports.safeJSON = safeJSON;
var _crypto = _interopRequireDefault(require("crypto"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function getNonce() {
  return _crypto.default.randomBytes(16).toString('base64').replace(/[^a-zA-Z0-9]+/g, '');
}
function getQuery(req) {
  const result = {};
  for (const key of Object.keys(req.query)) {
    const val = req.query[key];
    if (key && val && typeof val === 'string') {
      result[key] = val;
    }
  }
  return result;
}
function buildCSP(obj) {
  return Object.keys(obj).map(key => {
    return `${key} ${obj[key]};`;
  }).join('');
}
const TIME_OFFSET = 1564500000;
function normalizeTimestamp(timestamp) {
  return Math.floor(timestamp / 1000) - TIME_OFFSET;
}
function getTimestamp() {
  return normalizeTimestamp(Date.now());
}
function isIE(req) {
  const userAgent = req.get('user-agent');
  return Boolean(userAgent && /Edge|MSIE|rv:10|rv:11/i.test(userAgent));
}
function safeJSON(...args) {
  return JSON.stringify.apply(null, arguments).replace(/</g, '\\u003C').replace(/>/g, '\\u003E');
}
